src/v/cluster/self_test_frontend.cc

```diff
@@ -0,0 +1,263 @@
+}
+
+ss::future<self_test_frontend::local_test_state>
+self_test_frontend::local_invoke::stop_test() {
+    co_return enrich(co_await _self_be.local().stop_test());
+}
+
+ss::future<self_test_frontend::local_test_state>
+self_test_frontend::local_invoke::get_status() {
+    co_return enrich(_self_be.local().get_status());
+}
+
+/// Implementation of methods in self_test_frontend::remote_invoke
+
+ss::future<self_test_frontend::local_test_state>
+self_test_frontend::remote_invoke::start_test(start_test_request req) {
+    auto v = co_await send_self_test_rpc(
+      [req](self_test_rpc_client_protocol c) mutable {
+          return c.start_test(
+            std::move(req), rpc::client_opts(raft::clock_type::now() + 5s));
+      });
+    co_return enrich(v);
+}
+
+ss::future<self_test_frontend::local_test_state>
+self_test_frontend::remote_invoke::stop_test() {
+    auto v = co_await send_self_test_rpc([](self_test_rpc_client_protocol c) {
+        /// Bigger timeout to wait for executing jobs to shutdown
+        return c.stop_test(
+          empty_request{}, rpc::client_opts(raft::clock_type::now() + 30s));
+    });
+    co_return enrich(v);
+}
+
+ss::future<self_test_frontend::local_test_state>
+self_test_frontend::remote_invoke::get_status() {
+    auto v = co_await send_self_test_rpc([](self_test_rpc_client_protocol c) {
+        return c.get_status(
+          empty_request{}, rpc::client_opts(raft::clock_type::now() + 5s));
+    });
+    co_return enrich(v);
+}
+
+/// Implementation of methods in self_test_frontend
+
+self_test_frontend::self_test_frontend(
+  model::node_id self,
+  ss::sharded<cluster::members_table>& members,
+  ss::sharded<cluster::self_test_backend>& self_be,
+  ss::sharded<rpc::connection_cache>& connections)
+  : _self(self)
+  , _state({})
+  , _previous_successful_run({})
+  , _members(members)
+  , _self_be(self_be)
+  , _connections(connections) {}
+
+ss::future<> self_test_frontend::start() { return ss::now(); }
+
+ss::future<> self_test_frontend::stop() { co_await _gate.close(); }
+
+ss::future<uuid_t>
+self_test_frontend::start_test(diskcheck_opts dto, netcheck_opts nto) {
+    auto gate_holder = _gate.hold();
+    const auto stopped_results = co_await stop_test();
+    if (!stopped_results.finished()) {
+        /// hmm what to do? maybe throw? continue with less participants
+        co_return uuid_t{}; // TODO: fix
+    }
+
+    /// Invoke command to start test on all nodes, using the same test id
+    const auto id = uuid_t::create();
+    start_test_request req{.id = id, .dto = dto, .nto = nto};
+    co_await invoke_on_all_nodes(
+      [req](auto handle) { return handle->start_test(req); });
+    /// TODO: Throw if something bad occurred
+    co_return id;
+}
+
+ss::future<self_test_frontend::global_test_state>
+self_test_frontend::stop_test() {
+    auto gate_holder = _gate.hold();
+    int8_t retries = 3;
+    auto test_state = co_await status();
```

Comment:
I get the feeling we'd gain a lot of code clarity by explicitly keeping track of the state of the most recently run test. For instance, a `status()` call would be able to immediately return if the test were already finished rather than sending out RPCs; likewise, a `stop_test()` call on a test that has already completed would avoid similar work.

Reply:
Ah i remember why I didn't do that now, because in the case leadership changes then previous results wouldn't be cached on new leader. Asking nodes for their results via one RPC to each isn't such a big deal.

Reply:
I can probably do something a little more clever like checking if previous results is null or not before making that call, changes incoming..